src/walker.ts

```diff
@@ -0,0 +1,62 @@
+import { PotentialStep, Pipeline, Conditional, Command, CommandStep } from '.';
+import { getAndCacheDependency } from './conditional';
+
+export type Entity = Pipeline | PotentialStep | Command;
+
+async function walkEntity<T extends Entity>(
+  entity: T,
+  nodeCache: Map<any, any>,
+  mutator?: (entity: Entity) => any,
+): Promise<T> {
+  if (nodeCache.has(entity)) {
+    return nodeCache.get(entity);
+  }
+
+  const cache = new Map();
+
+  if (entity instanceof Conditional) {
+    if (await entity.accept()) {
+      entity = <any>await getAndCacheDependency(cache, entity);
+    } else {
+      return entity;
+    }
+  }
+
+  if (entity instanceof Pipeline) {
+    const steps = entity.steps;
+    steps.forEach(async (step, idx) => {
```

Comment:
async functions in `.forEach` might not behave the way you want them to. I think a standard `for` loop might be better here, making the walk block until the async mutation is done.